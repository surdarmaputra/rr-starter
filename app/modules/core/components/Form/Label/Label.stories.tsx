import type { Meta, StoryObj } from '@storybook/react';

import { Label, type LabelProps } from './Label';

type Story = StoryObj<typeof Label>;

const meta: Meta = {
  component: Label,
  title: 'Core/Components/Form/Label',
};

export default meta;

export const Default: Story = {
  render: (args: LabelProps) => <Label {...args}>Sample label</Label>,
};
