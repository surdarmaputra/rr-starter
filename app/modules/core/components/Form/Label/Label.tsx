import * as LabelPrimitive from '@radix-ui/react-label';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

const labelVariants = cva(
  'text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-letter-body',
);

export interface LabelProps
  extends React.ComponentPropsWithoutRef<typeof LabelPrimitive.Root>,
    VariantProps<typeof labelVariants> {
  secondaryText?: React.ReactNode;
}

const Label = React.forwardRef<
  React.ComponentRef<typeof LabelPrimitive.Root>,
  LabelProps
>(({ children, className, secondaryText, ...props }, ref) => (
  <LabelPrimitive.Root
    className={cn(labelVariants(), className)}
    ref={ref}
    {...props}
  >
    {children}
    {Boolean(secondaryText) && (
      <span className="ml-1.5 font-normal text-letter-caption">
        {secondaryText}
      </span>
    )}
  </LabelPrimitive.Root>
));
Label.displayName = LabelPrimitive.Root.displayName;

export { Label };
