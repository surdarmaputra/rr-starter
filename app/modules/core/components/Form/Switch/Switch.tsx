import * as SwitchPrimitives from '@radix-ui/react-switch';
import * as React from 'react';

import { mergeClassNames } from '~/modules/core/libs/utils/ui';

export type SwitchProps = React.ComponentPropsWithoutRef<
  typeof SwitchPrimitives.Root
>;

const Switch = React.forwardRef<
  React.ComponentRef<typeof SwitchPrimitives.Root>,
  SwitchProps
>(({ className, ...props }, ref) => (
  <SwitchPrimitives.Root
    className={mergeClassNames(
      'peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full',
      'border-2 border-transparent transition-colors',
      'focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 focus-visible:ring-offset-white',
      'disabled:cursor-not-allowed disabled:opacity-50',
      'data-[state=checked]:bg-primary data-[state=unchecked]:bg-primary-disabled dark:data-[state=unchecked]:bg-accent',
      className,
    )}
    {...props}
    ref={ref}
  >
    <SwitchPrimitives.Thumb
      className={mergeClassNames(
        'pointer-events-none block h-5 w-5 rounded-full bg-white shadow-lg dark:bg-black',
        'ring-0 transition-transform',
        'data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0',
      )}
    />
  </SwitchPrimitives.Root>
));
Switch.displayName = SwitchPrimitives.Root.displayName;

export { Switch };
