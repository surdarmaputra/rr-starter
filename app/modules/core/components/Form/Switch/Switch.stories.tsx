import type { Meta, StoryObj } from '@storybook/react';

import { Label } from '../Label/Label';
import { Switch, type SwitchProps } from './Switch';

type Story = StoryObj<typeof Switch>;

const meta: Meta = {
  component: Switch,
  title: 'Core/Components/Form/Switch',
};

export default meta;

export const Default: Story = {
  render: (args: SwitchProps) => (
    <div className="flex items-center space-x-2">
      <Switch id="airplane-mode" {...args} />
      <Label htmlFor="airplane-mode">Airplane Mode</Label>
    </div>
  ),
};
