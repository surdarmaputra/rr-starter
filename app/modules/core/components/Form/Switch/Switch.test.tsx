import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';

import { Label } from '../Label/Label';
import { Switch, type SwitchProps } from './Switch';

describe('Switch', () => {
  function renderComponent(props?: SwitchProps) {
    return render(
      <div className="flex items-center space-x-2">
        <Switch id="airplane-mode" {...props} />
        <Label htmlFor="airplane-mode">Airplane Mode</Label>
      </div>,
    );
  }

  test('triggers onCheckedChange', async () => {
    const handleCheckedChange = vi.fn();
    renderComponent({ onCheckedChange: handleCheckedChange });

    const switchToggle = screen.getByRole('switch', { name: /Airplane Mode/ });
    fireEvent.click(switchToggle);
    await waitFor(() => expect(handleCheckedChange).toHaveBeenCalledTimes(1));
    expect(handleCheckedChange).toHaveBeenCalledWith(true);

    fireEvent.click(switchToggle);
    await waitFor(() => expect(handleCheckedChange).toHaveBeenCalledTimes(2));
    expect(handleCheckedChange).toHaveBeenCalledWith(false);
  });

  test('renders snapshot', async () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
