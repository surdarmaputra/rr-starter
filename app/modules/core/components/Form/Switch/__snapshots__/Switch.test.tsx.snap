// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch > renders snapshot 1`] = `
<div>
  <div
    class="flex items-center space-x-2"
  >
    <button
      aria-checked="false"
      class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 focus-visible:ring-offset-white disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-primary-disabled dark:data-[state=unchecked]:bg-accent"
      data-state="unchecked"
      id="airplane-mode"
      role="switch"
      type="button"
      value="on"
    >
      <span
        class="pointer-events-none block h-5 w-5 rounded-full bg-white shadow-lg dark:bg-black ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
        data-state="unchecked"
      />
    </button>
    <label
      class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-letter-body"
      for="airplane-mode"
    >
      Airplane Mode
    </label>
  </div>
</div>
`;
